File: Spark/src/main/scala/SparkCassandra.scala
import com.datastax.driver.core.Session
import com.datastax.driver.core.exceptions.SyntaxError
import com.datastax.spark.connector.cql.CassandraConnector
import org.apache.spark.sql.{SaveMode, SparkSession}
import org.slf4j.{Logger, LoggerFactory}

/**
 *
 */
class SparkCassandra {

  val logger: Logger = LoggerFactory.getLogger(SparkCassandra.getClass)

  /**
   *
   * @param session
   * @return
   */
  private def createKeySpaceIfExists(session: Session): Boolean = {
    val result = session.execute("create keyspace if not exists movieKeySpace with replication = " +
      "{ 'class' : 'SimpleStrategy', 'replication_factor':3}")
    result.wasApplied()
  }

  /**
   *
   * @param session
   * @return
   */
  private def createAvgRatingsTable(session: Session): Boolean = {
    try {
      val result = session.execute("create table if not exists " +
        "moviekeyspace.movies_avg_ratings (movie_id int primary key, " +
        " movie_title varchar," +
        "avg_rating double)")
      result.wasApplied()
    } catch {
      case ex: SyntaxError => logger.error("Error occurred -> " + ex)
        session.close()
        return false
      case ex: Exception => logger.error("Error occurred ->" + ex)
        session.close()
        return false
    }
  }

  /**
   *
   * @param sparkSession
   * @param session
   */
  private def populateAvgRatingsTable(sparkSession: SparkSession,
                                      session: Session): Unit = {

    logger.info("Reading the csv file from hadoop")

    sparkSession.
      read.
      format("csv").
      option("inferSchema", "true").
      option("header", "true").
      load("hdfs://localhost:9000/user/hadoop/data/ratings.csv").
      createOrReplaceTempView("movies_ratings")

    logger.info("View movies_ratings successfully created")

    sparkSession.
      read.
      format("csv").
      option("inferSchema", "true").
      option("header", "true").
      load("hdfs://localhost:9000/user/hadoop/data/movies.csv").
      createOrReplaceTempView("movies")

    logger.info("View movies successfully created")

    sparkSession.sql("" +
      "SELECT movieid " +
      "FROM movies_ratings " +
      "GROUP BY movieid " +
      "HAVING count(1)> 10").createOrReplaceTempView("valid_movies")

    logger.info("View valid_movies successfully created")

    val resultAvgMovieRatingDF = sparkSession.sql("select v.movieid as movie_id, " +
      "m.title movie_title ,avg(r.rating) as avg_rating " +
      "from movies m INNER JOIN movies_ratings r " +
      " on (m.movieid = r.movieid)" +
      "INNER JOIN valid_movies v" +
      " on (r.movieid = v.movieid)" +
      "group by v.movieid, m.title order by avg(r.rating)")

    logger.info("Dataframe resultAvgMovieRatingDF successfully created")

    logger.info("Trying to populate the movies_avg_ratings in Cassandra")
    try {
      resultAvgMovieRatingDF.
        write.
        format("org.apache.spark.sql.cassandra").
        options(Map("table" -> "movies_avg_ratings",
          "keyspace" -> "moviekeyspace",
          "confirm.truncate" -> "true")).
        mode(saveMode = SaveMode.Overwrite).
        save()

      logger.info("Successfully populated the movies_avg_ratings in Cassandra")

      logger.info("Trying to verify the existance of Cassandra table after all the above operations")

      try {
        val createDDL =
          """ CREATE TEMPORARY VIEW movies_avg_ratings
        USING org.apache.spark.sql.cassandra
            |OPTIONS ( table "movies_avg_ratings",
            |keyspace "moviekeyspace",
            |pushdown "true")""".stripMargin

        sparkSession.sql(createDDL)
        val countDF = sparkSession.sql("select count(1) from movies_avg_ratings")
        logger.info("Table counts = " ++ countDF.collect().mkString(""))

        logger.info("Cassandra table has been sucessfully created and populated using Spark")
      } catch {
        case ex: Exception => logger.error("Error occurred while trying to fetch table from cassandra => " + ex)
          session.close()
          sparkSession.close()
      }

    } catch {
      case ex: Exception => logger.error("Exception -> " + ex)
        session.close()
        sparkSession.close()
    }
  }

  /**
   * Creating a cassandra session for executing the SQL stmts inside cassandra key space
   *
   * @param sparkSession - SparkSession
   * @return - DataStax Cassandra Session used for executing stmts
   */
  private def getCassandraConnection(sparkSession: SparkSession): Session = {
    val cassandraSession = CassandraConnector.apply(sparkSession.sparkContext.getConf)
    cassandraSession.openSession()
  }

  /**
   *
   */
  def init(): Unit = {
    val sparkSession = SparkSession.
      builder().
      master("local[*]").
      appName("SparkCassandra").
      config("spark.cassandra.connection.host", "localhost").
      config("spark.cassandra.connection.port", "9042").
      config("spark.cassandra.output.consistency.level", "LOCAL_ONE").
      config("spark.cassandra.input.consistency.level", "LOCAL_ONE").
      getOrCreate()

    logger.info("Trying to get the cassandra connection")

    val cassandraSession: Session = getCassandraConnection(sparkSession)

    logger.info("Got the cassandra connection")

    logger.info("Trying to create the keyspace if it doesnt exist")

    if (!createKeySpaceIfExists(cassandraSession))

      logger.error("Something went wrong... Please check with administrator")

    else {
      // create cassandra table
      logger.info("Trying to create table if it doesnt exist")

      if (!createAvgRatingsTable(cassandraSession)) {

        logger.error("Something went wrong... Please check with administrator")

      } else {

        logger.info("Populating the table")

        populateAvgRatingsTable(sparkSession, session = cassandraSession)

        logger.info("Processing complete")
      }
    }
    logger.info("Closing the cassendra session")

    cassandraSession.close()

    logger.info("Cassandra Session closed")

    sparkSession.close()

    logger.info("Spark session closed")
  }

}

object SparkCassandra extends App{

  val sparkCassandra = new SparkCassandra
  sparkCassandra.init()
}

File: Spark/spark-cassandra/python/SparkCassandra.py
import cassandra
import pyspark
from cassandra.cluster import Cluster
from pyspark.sql import SparkSession
from pyspark import SparkContext
from pyspark.sql.functions import *


def get_cassandra_connection(host_name):
    cluster  = Cluster([host_name])
    return cluster.connect()


def get_spark_session(host, port):
    return (SparkSession.\
                builder.\
                master("local[*]").\
                appName("SparkCassandra").\
                config("spark.cassandra.connection.host", host).\
                config("spark.cassandra.connection.port", port).\
                config("spark.cassandra.output.consistency.level", "LOCAL_ONE").\
                config("spark.cassandra.input.consistency.level", "LOCAL_ONE").\
                getOrCreate())


def create_keyspace(session, keyspace, replication_options = None):
    
    replication_options = replication_options or \
                          "{'class': 'SimpleStrategy', 'replication_factor': 3}"
    
    result = session.execute(""" 
            CREATE KEYSPACE IF NOT EXISTS {} 
            WITH REPLICATION = {} """.format(keyspace, replication_options))


def create_avg_rating_by_movies_table(session, keyspace):
    session.execute("""
                CREATE TABLE IF NOT EXISTS {}.avg_rating_by_occupation 
                    (occupation TEXT PRIMARY KEY, 
                     rating_count DOUBLE);
        """.format(keyspace))


def create_avg_rating_by_movies_table(session, keyspace):
    session.execute("""
            CREATE TABLE IF NOT EXISTS {}.avg_rating_by_movies 
                (movie_id INT PRIMARY KEY, 
                title TEXT, 
                release_dt TEXT,
                rating_count DOUBLE, 
                avg_rating DOUBLE );
        """.format(keyspace))

    
def get_ratings_df(spark):
    return (spark.sparkContext.\
        textFile("file:///home/hadoop/tutorial/HadoopMaterials/ml-100k/u.data").\
        map(lambda f : f.split("\t") ).\
        toDF().\
        selectExpr("cast(_1 as int) as user_id",\
        "cast(_2 as int) as movie_id",\
        "cast(_3 as double) as rating",\
        "cast(_4 as bigint) as timestamp"))


def get_users_df(spark):
    return (spark.sparkContext.\
        textFile("file:///home/hadoop/tutorial/HadoopMaterials/ml-100k/u.user").\
        map(lambda f : f.split("|") ).\
        toDF().\
        selectExpr("cast(_1 as int) as user_id",\
        "cast(_2 as int) as age",\
        "cast(_3 as string) as gender",\
        "cast(_4 as string) as occupation",\
        "cast(_5 as string) as zipcode"))


def get_movies_df(spark):    
    return (spark.sparkContext.\
        textFile("file:///home/hadoop/tutorial/HadoopMaterials/ml-100k/u.item").\
        map(lambda f : f.split("|") ).\
        toDF().\
        selectExpr("cast(_1 as int) as movie_id",\
        "cast(_2 as string) as title",\
        "cast(_3 as string) as release_dt"))


def get_avg_ratings_df(ratingsDF):
    return (ratingsDF.\
        selectExpr("movie_id as ratings_movie_id","rating").\
        groupBy("ratings_movie_id").\
        agg(count("rating").alias("rating_count"), \
        avg("rating").alias("avg_rating")).\
        filter("rating_count > 10"))
    

def get_avg_user_ratings_df(ratingsDF):
    return (ratingsDF.selectExpr("user_id as ratings_user_id", "rating").\
        groupBy("ratings_user_id", "rating").\
        agg(count("rating").alias("rating_count")).\
        filter("rating_count > 10").\
        drop("rating_count"))


def prepare_avg_rating_by_movies_dataset(avg_ratings_df, moviesDF):   
    join_expression = avg_ratings_df["ratings_movie_id"] == moviesDF["movie_id"]

    join_type = "inner"

    avg_rating_by_movies_dataset = \
    moviesDF.\
    join(avg_ratings_df,join_expression,join_type).\
    drop("ratings_movie_id").\
    sort(desc("avg_rating"))
    
    avg_rating_by_movies_dataset.\
    write.\
        format("org.apache.spark.sql.cassandra").\
        options(keyspace="movielens",\
                table="avg_rating_by_movies").\
        option("confirm.truncate","true").\
        mode("overwrite").\
        save()


def prepare_avg_rating_by_occupation(avg_user_ratings_df, usersDF):
    join_expression = avg_user_ratings_df["ratings_user_id"] == usersDF["user_id"]

    join_type = "inner"

    avg_rating_by_occupation_dataset = \
    usersDF.join(avg_user_ratings_df,join_expression,join_type).\
    drop("ratings_user_id","age","zipcode","user_id","gender").\
    groupBy("occupation").\
    agg(count("rating").alias("rating_count")).\
    sort(desc("rating_count"))
    
    avg_rating_by_occupation_dataset.\
    write.\
        format("org.apache.spark.sql.cassandra").\
        options(keyspace="movielens",\
                table="avg_rating_by_occupation").\
        option("confirm.truncate","true").\
        mode("overwrite").\
        save()


def read_get_avg_rating_by_movies_from_db(spark):
    return (spark.read.format("org.apache.spark.sql.cassandra").\
    options(keyspace="movielens",\
            table="avg_rating_by_movies").\
    load())



def read_avg_rating_by_occupation_from_db(spark):
    return (spark.read.format("org.apache.spark.sql.cassandra").\
    options(keyspace="movielens",\
            table="avg_rating_by_occupation").\
    load())


def init(host_name, port, keyspace):
    
    print("host = {}: port = {} : keyspace = {}".format(host_name, port, keyspace))

    spark = get_spark_session(host_name, port)
    if (spark == None):
        print("sparkSession is null")
        exit
    else:

        # Setting the logging level
        spark.sparkContext.setLogLevel("ERROR")
        
        # Reading from source files
        ratingsDF = get_ratings_df(spark)
        usersDF = get_users_df(spark)
        moviesDF = get_movies_df(spark)

        ## Prepating the data for inserting..
        avg_ratings_df = get_avg_ratings_df(ratingsDF)
        #avg_ratings_df.show()
        
        avg_user_ratings_df = get_avg_user_ratings_df(ratingsDF)
        #avg_user_ratings_df.show()
        
        # Prepare cassandra session.
        session = get_cassandra_connection(host_name)
        
        # Create the keyspace
        create_keyspace(session, keyspace)
        
        session.set_keyspace(keyspace)
                
        # Create the avg_rating_by_movies table in the keyspace
        create_avg_rating_by_movies_table(session, keyspace)
        
        # Create the avg_rating_by_movies table in the keyspace
        create_avg_rating_by_movies_table(session, keyspace)

        ## Writing into the mongodb database
        prepare_avg_rating_by_movies_dataset(avg_ratings_df, moviesDF)
        prepare_avg_rating_by_occupation(avg_user_ratings_df, usersDF)
        
        # Reading the "avg_rating_by_movies" inserted data from database 
        # and displaying on the console
        read_get_avg_rating_by_movies_from_db(spark).drop("_id").sort(desc("avg_rating")).show()

        # Reading the "avg_rating_by_occupation" inserted data from database 
        # and displaying on the console
        read_avg_rating_by_occupation_from_db(spark).drop("_id").sort(desc("rating_count")).show()
        
        # Closing the cassandra session
        session.shutdown()

        # Stopping the spark session
        spark.stop()

import sys

def main():
    
    print(len(sys.argv))

    if (len(sys.argv) != 4):
        print("Please pass the appropriate arguments to the program..\n")
        print("\t Usage >> \n")
        print("\t\t SparkCassandra <cassandra host name>> <<cassandra port>> <<cassandra keyspace>>\n\n")

    else:

        host_name = sys.argv[1].strip()
        port      = sys.argv[2].strip()
        keyspace  = sys.argv[3].strip()
        
        if host_name is None or host_name == "":
            host_name = "127.0.0.1"
        
        if port is None or port == "":
            port = "9042"
            
        if keyspace is None or keyspace == "":
            keyspace = "movielens"       
        
        init(host_name, 
             port,
             keyspace)


if __name__ == "__main__":
    main()
